src/api/ucc.h

```diff
@@ -0,0 +1,709 @@
+ */
+
+ucc_status_t ucc_team_create_from_parent(uint64_t my_ep, bool included,
+                                             ucc_team_h parent_team, 
+                                             ucc_team_h *new_ucc_team);
+
+/**
+ * @ingroup UCC_TEAM
+ *
+ * @ref ucc_team_get_size routine queries the size of the team. It reflects the
+ * number of unique endpoints in the team.
+ *
+ * @param [in]   ucc_team  Team handle
+ * @param [out]  size      The size of team as number of endpoints
+ *
+ * @return Error code as defined by @ref ucc_status_t
+ */
+
+ucc_status_t ucc_team_get_size(ucc_team_h ucc_team, uint32_t *size);
+
+
+/**
+ * @ingroup UCC_TEAM
+ *
+ * @ref ucc_team_get_my_ep routine queries and returns the endpoint of the
+ * participant invoking the interface.
+ *
+ * @param [out]  ep          Endpoint of the participant calling the routine
+ * @param [in]   ucc_team    Team handle
+ *
+ * @return Error code as defined by @ref ucc_status_t
+ */
+
+ucc_status_t ucc_team_get_my_ep(ucc_team_h ucc_team, uint64_t *ep);
+
+/**
+ * @ingroup UCC_TEAM
+ *
+ * @ref ucc_team_my_ep routine queries and returns all endpoints of all
+ * participants in the team.
+ *
+ * @param [out]     ep          List of endpoints
+ * @param [out]     num_eps     Number of endpoints
+ * @param [in]      ucc_team    Team handle
+ *
+ * @return Error code as defined by @ref ucc_status_t
+ */
+
+ucc_status_t ucc_team_get_all_eps(ucc_team_h ucc_team, uint64_t **ep, uint64_t
+                                  *num_eps);
+
+
+/**
+ * *************************************************************
+ *                   Collectives Section 
+ * *************************************************************
+ */
+typedef enum {
+    UCC_DT_INT8 = 0,
+    UCC_DT_INT16,
+    UCC_DT_INT32,
+    UCC_DT_INT64,
+    UCC_DT_INT128,
+    UCC_DT_UINT8,
+    UCC_DT_UINT16,
+    UCC_DT_UINT32,
+    UCC_DT_UINT64,
+    UCC_DT_UINT128,
+    UCC_DT_FLOAT16,
+    UCC_DT_FLOAT32,
+    UCC_DT_FLOAT64,
+    UCC_DT_USERDEFINED,
+    UCC_DT_OPAQUE
+} ucc_datatype_t;
+
+typedef struct ucc_coll_buffer_info {
+    uint64_t        mask;
+    void            *src_buffer;
+    ucc_count_t     *src_counts;
+    ucc_aint_t      *src_displacements;
+    void            *dst_buffer;
+    ucc_count_t     *dst_counts;
+    ucc_aint_t      *dst_displacements;
+    ucc_datatype_t  src_datatype;
+    ucc_datatype_t  dst_datatype;
+    uint64_t        flags;
+} ucc_coll_buffer_info_t;
+
+typedef struct ucc_reduction_info {
+    ucc_datatype_t      dt;
+    ucc_reduction_op_t  op;
+    size_t              count;
+} ucc_reduction_info_t;
+
+typedef enum {
+    UCC_ERR_TYPE_LOCAL=0,
+    UCC_ERR_TYPE_GLOBAL=1
+} ucc_error_type_t;
+
+typedef uint16_t ucc_coll_id_t ;
+
+typedef struct ucc_coll_ext_op_args {
+    ucc_coll_buffer_info_t      buffer_info;
+    ucc_reduction_info_t        reduction_info;
+    ucc_error_type_t            error_type;
+    ucc_coll_id_t               tag;
```

Comment:
MPI collectives API doesn't have tags - do you mean this will be an optional feature, which will behave like P2P tags? If so, is it "always on"? (if always on, sounds like we should have a way to turn it off, to save cycles)

Reply:
it is definitely optional and off by default. Supposed to be used for "unordered" collectives (commonly used in ML frameworks).

Reply:
Sounds good, but In that case - it should probably be indicated by a flag or something...

Reply:
I'm not sure, I understand. The team is created with option of ordered or unordered collectives. Then, the tag is a required field for all unordered collectives, and it is not required for all collectives on "ordered" team. 

Let's touch upon this in the next call.